File: notification-service/src/main/java/pesco/notification_service/configurations/RabbitMQConfig.java
package pesco.notification_service.configurations;

import org.springframework.amqp.core.*;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class RabbitMQConfig {

    // Constants for exchange and queues
    public static final String WALLET_EXCHANGE = "wallet.notifications";
    public static final String AUTH_EXCHANGE = "auth.notifications";

    // Auth & User Queues
    public static final String ACCOUNT_VERIFICATION_QUEUE = "account.verification";
    public static final String USER_OTP_QUEUE = "user.otp";
    public static final String RESET_PASSWORD_QUEUE = "passsword.reset";

    // Wallet Queues
    public static final String CREDIT_WALLET_QUEUE = "credit.wallet";
    public static final String DEBIT_WALLET_QUEUE = "debit.wallet";
    public static final String DEPOSIT_WALLET_QUEUE = "deposit.wallet";
    public static final String MAINTENANCE_DEDUCTION_QUEUE = "maintenance.service";

    // Routing keys
    public static final String ROUTING_KEY_ACCOUNT_VERIFICATION = "auth.account";
    public static final String ROUTING_KEY_ACCOUNT_USER_OTP = "auth.user";
    public static final String ROUTING_KEY_RESET_PASSWORD = "auth.passsword";

    public static final String ROUTING_KEY_CREDIT_WALLET = "wallet.credit";
    public static final String ROUTING_KEY_DEBIT_WALLET = "wallet.debit";
    public static final String ROUTING_KEY_DEPOSIT_WALLET = "wallet.deposit";
    public static final String ROUTING_KEY_WALLET_MAINTENANCE_SERVICE_DEDUCTION = "wallet.deduction-fee";

    // Declare the exchange
    @Bean
    public TopicExchange accountExchange() {
        return new TopicExchange(AUTH_EXCHANGE);
    }

    @Bean
    public TopicExchange walletExchange() {
        return new TopicExchange(WALLET_EXCHANGE);
    }

    // Declare the queues
    @Bean
    public Queue accountVerificationQueue() {
        return new Queue(ACCOUNT_VERIFICATION_QUEUE);
    }

    @Bean
    public Queue userOTPQueue() {
        return new Queue(USER_OTP_QUEUE);
    }

    @Bean
    public Queue resetPasswordQueue() {
        return new Queue(RESET_PASSWORD_QUEUE);
    }

    @Bean
    public Queue creditWalletQueue() {
        return new Queue(CREDIT_WALLET_QUEUE);
    }

    @Bean
    public Queue debitWalletQueue() {
        return new Queue(DEBIT_WALLET_QUEUE);
    }

    @Bean
    public Queue depositWalletQueue() {
        return new Queue(DEPOSIT_WALLET_QUEUE);
    }
    
    @Bean
    public Queue walletMaintenanceDeductionQueue() {
        return new Queue(MAINTENANCE_DEDUCTION_QUEUE);
    }

    // Bindings
    @Bean
    public Binding accountVerificationBinding(Queue accountVerificationQueue, TopicExchange accountExchange) {
        return BindingBuilder.bind(accountVerificationQueue).to(accountExchange).with(ROUTING_KEY_ACCOUNT_VERIFICATION);
    }

    @Bean
    public Binding userOTPBinding(Queue userOTPQueue, TopicExchange accountExchange) {
        return BindingBuilder.bind(userOTPQueue).to(accountExchange).with(ROUTING_KEY_ACCOUNT_USER_OTP);
    }

    @Bean
    public Binding resetPasswordBinding(Queue resetPasswordQueue, TopicExchange accountExchange) {
        return BindingBuilder.bind(resetPasswordQueue).to(accountExchange).with(ROUTING_KEY_RESET_PASSWORD);
    }

    @Bean
    public Binding creditWalletBinding(Queue creditWalletQueue, TopicExchange walletExchange) {
        return BindingBuilder.bind(creditWalletQueue).to(walletExchange).with(ROUTING_KEY_CREDIT_WALLET);
    }

    @Bean
    public Binding debitWalletBinding(Queue debitWalletQueue, TopicExchange walletExchange) {
        return BindingBuilder.bind(debitWalletQueue).to(walletExchange).with(ROUTING_KEY_DEBIT_WALLET);
    }

    @Bean
    public Binding depositWalletBinding(Queue depositWalletQueue, TopicExchange walletExchange) {
        return BindingBuilder.bind(depositWalletQueue).to(walletExchange).with(ROUTING_KEY_DEPOSIT_WALLET);
    }

    @Bean
    public Binding walletMaintenanceBinding(Queue walletMaintenanceDeductionQueue, TopicExchange walletExchange) {
        return BindingBuilder.bind(walletMaintenanceDeductionQueue).to(walletExchange).with(ROUTING_KEY_WALLET_MAINTENANCE_SERVICE_DEDUCTION);
    }

    // Configure RabbitTemplate with Jackson2JsonMessageConverter for JSON serialization
    @Bean
    public Jackson2JsonMessageConverter messageConverter() {
        return new Jackson2JsonMessageConverter();
    }

    @Bean
    public RabbitTemplate rabbitTemplate(ConnectionFactory connectionFactory) {
        RabbitTemplate rabbitTemplate = new RabbitTemplate(connectionFactory);
        rabbitTemplate.setMessageConverter(messageConverter());
        return rabbitTemplate;
    }
}

File: app/rabbitmq_config.py
import os
import pika

# Wallet Exchange & Queues
WALLET_EXCHANGE = "wallet.notifications"
CREDIT_WALLET_QUEUE = "credit.wallet"
DEBIT_WALLET_QUEUE = "debit.wallet"
DEPOSIT_WALLET_QUEUE = "deposit.wallet"
MAINTENANCE_DEDUCTION_QUEUE = "maintenance.service"

ROUTING_KEY_CREDIT_WALLET = "wallet.credit"
ROUTING_KEY_DEBIT_WALLET = "wallet.debit"
ROUTING_KEY_DEPOSIT_WALLET = "wallet.deposit"
ROUTING_KEY_WALLET_MAINTENANCE = "wallet.deduction-fee"

# Authentication Exchange & Queues
AUTH_EXCHANGE = "auth.notifications"
EMAIL_VERIFICATION_QUEUE = "email.verification"
EMAIL_OTP_QUEUE = "email.otp"
EMAIL_RESET_PASSWORD_QUEUE = "email.reset-password"

ROUTING_KEY_EMAIL_VERIFICATION = "email.verification"
ROUTING_KEY_EMAIL_OTP = "email.otp"
ROUTING_KEY_EMAIL_RESET_PASSWORD = "email.reset-password"

def setup_rabbitmq():
    """Initialize RabbitMQ exchanges, queues, and bindings."""
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host=os.getenv("RABBITMQ_HOST", "rabbitmq"))
    )
    channel = connection.channel()

    # Declare Exchanges
    channel.exchange_declare(exchange=WALLET_EXCHANGE, exchange_type="topic", durable=True)
    channel.exchange_declare(exchange=AUTH_EXCHANGE, exchange_type="direct", durable=True)

    # Declare Wallet Queues
    channel.queue_declare(queue=CREDIT_WALLET_QUEUE, durable=True)
    channel.queue_declare(queue=DEBIT_WALLET_QUEUE, durable=True)
    channel.queue_declare(queue=DEPOSIT_WALLET_QUEUE, durable=True)
    channel.queue_declare(queue=MAINTENANCE_DEDUCTION_QUEUE, durable=True)

    # Declare Authentication Queues
    channel.queue_declare(queue=EMAIL_VERIFICATION_QUEUE, durable=True)
    channel.queue_declare(queue=EMAIL_OTP_QUEUE, durable=True)
    channel.queue_declare(queue=EMAIL_RESET_PASSWORD_QUEUE, durable=True)

    # Bind Wallet Queues to Exchange
    channel.queue_bind(exchange=WALLET_EXCHANGE, queue=CREDIT_WALLET_QUEUE, routing_key=ROUTING_KEY_CREDIT_WALLET)
    channel.queue_bind(exchange=WALLET_EXCHANGE, queue=DEBIT_WALLET_QUEUE, routing_key=ROUTING_KEY_DEBIT_WALLET)
    channel.queue_bind(exchange=WALLET_EXCHANGE, queue=DEPOSIT_WALLET_QUEUE, routing_key=ROUTING_KEY_DEPOSIT_WALLET)
    channel.queue_bind(exchange=WALLET_EXCHANGE, queue=MAINTENANCE_DEDUCTION_QUEUE, routing_key=ROUTING_KEY_WALLET_MAINTENANCE)

    # Bind Authentication Queues to Exchange
    channel.queue_bind(exchange=AUTH_EXCHANGE, queue=EMAIL_VERIFICATION_QUEUE, routing_key=ROUTING_KEY_EMAIL_VERIFICATION)
    channel.queue_bind(exchange=AUTH_EXCHANGE, queue=EMAIL_OTP_QUEUE, routing_key=ROUTING_KEY_EMAIL_OTP)
    channel.queue_bind(exchange=AUTH_EXCHANGE, queue=EMAIL_RESET_PASSWORD_QUEUE, routing_key=ROUTING_KEY_EMAIL_RESET_PASSWORD)

    print(" [v] RabbitMQ Exchanges, Queues, and Bindings are set up.")

    connection.close()
